src/index.ts

```diff
@@ -151,61 +152,67 @@ async function runTests(
-            const version = typesVersions[i];
-            const versionPath = joinPaths(dirPath, `ts${version}`);
-            const versionIndexText = await readFile(joinPaths(versionPath, "index.d.ts"), "utf-8");
-            await testTypesVersion(
-                versionPath, version, getTsVersion(i + 1), isOlderVersion, dt, versionIndexText,
-                expectOnly, undefined, /*inTypesVersionDirectory*/ true);
-        }
-
-        function getTsVersion(i: number): TsVersion {
-            return i === typesVersions.length
-                ? TypeScriptVersion.latest
-                : assertDefined(TypeScriptVersion.previous(typesVersions[i]));
+        // For example, typesVersions of [3.2, 3.5, 3.6] will have
+        // associated ts3.2, ts3.5, ts3.6 directories, for
+        // <=3.2, <=3.5, <=3.6 respectively; the root level is for 3.7 and above.
+        // so this code needs to generate ranges [lowest-3.2, 3.3-3.5, 3.6-3.6, 3.7-latest]
+        const lows = [TypeScriptVersion.lowest, ...typesVersions.map(next)];
+        const his = [...typesVersions, TypeScriptVersion.latest];
+        assert.strictEqual(lows.length, his.length);
+        for (let i = 0; i < lows.length; i++) {
+            const low = maxVersion(minVersion, lows[i]);
+            const hi = his[i];
+            assert(
+                parseFloat(hi) >= parseFloat(low),
+                `'// Minimum TypeScript Version: ${minVersion}' in header skips ts${hi} folder.`);
+            const isLatest = hi === TypeScriptVersion.latest;
+            const versionPath = isLatest ? dirPath : joinPaths(dirPath, `ts${hi}`);
+            if (lows.length > 1) {
+                console.log("testing from", low, "to", hi, "in", versionPath);
+            }
+            await testTypesVersion(versionPath, low, hi, isOlderVersion, dt, expectOnly, undefined, isLatest);
         }
     }
 }
 
+function maxVersion(v1: TypeScriptVersion | undefined, v2: TypeScriptVersion | undefined): TypeScriptVersion;
+function maxVersion(v1: AllTypeScriptVersion | undefined, v2: AllTypeScriptVersion | undefined): AllTypeScriptVersion;
+function maxVersion(v1: AllTypeScriptVersion | undefined, v2: AllTypeScriptVersion | undefined) {
+    if (!v1) return v2;
+    if (!v2) return v1;
+    if (parseFloat(v1) >= parseFloat(v2)) return v1;
+    return v2;
+}
+
+function next(v: TypeScriptVersion): TypeScriptVersion {
+    const index = TypeScriptVersion.supported.indexOf(v);
+    assert.notStrictEqual(index, -1);
+    assert(index < TypeScriptVersion.supported.length);
+    return TypeScriptVersion.supported[index + 1];
+}
+
 async function testTypesVersion(
     dirPath: string,
-    lowVersion: TsVersion | undefined,
-    maxVersion: TsVersion,
+    lowVersion: TsVersion,
+    hiVersion: TsVersion,
     isOlderVersion: boolean,
     dt: boolean,
-    indexText: string,
     expectOnly: boolean,
     tsLocal: string | undefined,
-    inTypesVersionDirectory?: boolean,
+    isLatest: boolean,
```

Comment:
testTypesVersion is now called in many fewer places, so I

1. Shifted undefined checking out, to only places it applies.
2. Shifted min-version checking out, to only places it applies.
3. Made the last parameter required.